crates/turborepo-lib/src/graph/walker.rs

```diff
@@ -0,0 +1,257 @@
+                        // do some err handling with the send failure?
+                        if node_tx.send((node, callback_tx)).await.is_err() {
+                            // Receiving end of node channel has been closed/dropped
+                            // Since there's nothing the mark the node as being done
+                            // we act as if we have been canceled.
+                            trace!("Receiver was dropped before walk finished without calling cancel");
+                            return;
+                        }
+                        if callback_rx.await.is_err() {
+                            // If the caller drops the callback sender without signaling
+                            // that the node processing is finished we assume that it is finished.
+                            trace!("Callback sender was dropped without sending a finish signal")
+                        }
+                        // Send errors indicate that there are no receivers which
+                        // happens when this node has no dependents
+                        tx.send(()).ok();
+                    }
+                    _ = cancel_rx.changed() => {
+                        // If this future resolves this means that either:
+                        // - cancel was updated, this can only happen through
+                        //   the cancel method which only sets it to true
+                        // - the cancel sender was dropped which is also a sign that we should exit
+                    }
+                }
+            }));
+        }
+
+        // All senders should have been moved into their correct node task
+        debug_assert!(txs.is_empty());
+
+        Self {
+            cancel,
+            node_events: Some(node_rx),
+            join_handles,
+        }
+    }
+
+    /// Cancel the walk
+    /// Any nodes that are already in the queue to be emitted will still be
+    /// sent, but no new nodes will be sent.
+    pub fn cancel(&mut self) -> Result<(), watch::error::SendError<bool>> {
+        self.cancel.send(true)
+    }
+
+    /// Consumes the watcher and waits for all futures to be finished.
+    /// Primarily used after the walk is canceled to ensure all tasks
+    /// have been stopped.
+    pub async fn wait(self) -> Result<(), tokio::task::JoinError> {
+        let Self { join_handles, .. } = self;
+        // Wait for all of the handles to finish up
+        for handle in join_handles {
+            handle.await.ok();
+        }
+        Ok(())
+    }
+
+    /// Start walking the graph and return a channel that emits node
+    /// indices once all of the nodes dependencies have finished.
+    /// It is up to the caller to use the oneshot channel to mark
+    /// a node as being done.
+    pub fn walk(&mut self) -> mpsc::Receiver<WalkMessage<N>> {
+        self.node_events
+            .take()
+            .expect("walking graph with walker that has already been used")
```

Comment:
Could we make this function take `mut self` (and a type state pattern) instead to enforce that this walker can only be used once?